xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -0,0 +1,251 @@
+        ClusterWatcherImpl clusterWatcher;
+
+        @Override
+        public boolean isAvailable() {
+          return true;
+        }
+
+        @Override
+        public int getPriority() {
+          return 5;
+        }
+
+        /**
+         * A synthetic policy name identified by CDS config. The implementation detail doesn't
+         * matter.
+         */
+        @Override
+        public String getPolicyName() {
+          return "cds_policy__cluster_name_" + newCdsConfig.name;
+        }
+
+        @Override
+        public LoadBalancer newLoadBalancer(final Helper helper) {
+          return new LoadBalancer() {
+            @Override
+            public void handleNameResolutionError(Status error) {}
+
+            @Override
+            public void shutdown() {
+              clusterWatcher.edsBalancer.shutdown();
+            }
+
+            @Override
+            public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+              if (clusterWatcher == null) {
+                clusterWatcher = new ClusterWatcherImpl(helper, resolvedAddresses);
+                xdsClient.watchClusterData(newCdsConfig.name, clusterWatcher);
+                if (oldCdsConfig != null) {
+                  xdsClient.cancelClusterDataWatch(oldCdsConfig.name, oldClusterWatcher);
+                }
+                CdsLoadBalancer.this.clusterWatcher = clusterWatcher;
+              }
+            }
+          };
+        }
+      };
+
+      switchingLoadBalancer.switchTo(fixedCdsConfigBalancer);
+    }
+
+    switchingLoadBalancer.handleResolvedAddresses(resolvedAddresses);
+
+    // clusterWatcher is also updated after switchingLoadBalancer.handleResolvedAddresses
+    cdsConfig = newCdsConfig;
+  }
+
+  @Override
+  public void handleNameResolutionError(Status error) {
+    channelLogger.log(ChannelLogLevel.DEBUG, "Name resolution error: '%s'", error);
+  }
+
+  @Override
+  public boolean canHandleEmptyAddressListFromNameResolution() {
+    return true;
+  }
+
+  @Override
+  public void shutdown() {
+    channelLogger.log(ChannelLogLevel.DEBUG, "CDS load balancer is shutting down");
+
+    switchingLoadBalancer.shutdown();
+    if (xdsClientRef != null) {
+      xdsClientRef.returnObject(xdsClient);
+    }
+  }
+
+  private final class ClusterWatcherImpl implements ClusterWatcher {
+
+    final LoadBalancer edsBalancer;
+    final Helper helper;
+    final ResolvedAddresses resolvedAddresses;
+
+    ClusterWatcherImpl(Helper helper, ResolvedAddresses resolvedAddresses) {
+      this.helper = helper;
+      this.resolvedAddresses = resolvedAddresses;
+      this.edsBalancer = lbRegistry.getProvider("xds_experimental").newLoadBalancer(helper);
+    }
+
+    @Override
+    public void onClusterChanged(ClusterUpdate update) {
+      Preconditions.checkArgument(
+          update.getLbPolicy().equals("round_robin"),
+          "The lbPolicy in ClusterUpdate '%s' is not 'round_robin'", update);
```

Comment:
nit: error messages should be descriptive about the logic, not exposing variable names. Otherwise, if you changed variable names later, you have to change error message as well.

In this case, it's better to be "Only round robin in-cluster load balancing policy is supported."

Small thing, change it if you agree.

Reply:
Rephrased.